cms/models/placeholdermodel.py

```diff
@@ -498,7 +498,7 @@ def add_plugin(self, instance):
             self._shift_plugin_positions(
                 instance.language,
                 start=instance.position,
-                offset=last_position,
+                offset=last_position - instance.position + 2, # behind last_position plus one for the new
```

Comment:
I can't see the need for this logic, because the "last" plugin should be the end, if we are adding a plugin we need to shift the tree behind where we are inserting all of the way to the end.

Reply:
Like for the move between placeholders you have to both consider to be able to **move out the tail** (as you assume) but then also be able to **move the tail back in** after the new plugin has been added. For this you need one more spare position, since the gap has been reduced by the newly created plugin.

Moving the tail to the end amounts to an offset of `last_position - instance.position + 1`, then add `1` for the new plugin.

Consider your want to insert Plugin 0 at position 1 (positions in the first column):

```
1 Plugin 1
2 Plugin 2
```

First you move the tail out and put the new plugin into the database. Take as offset `last_position(2) - instance.position(1) + 2 = 3`:
```
1 Plugin 0
4 Plugin 1 
5 Plugin 2
```

Now you have the minimum position slots between `1` and `4` to move the tail back in (position 2 and up).

If you had just moved by `last_position(2)` the gap would not have been big enough to move back the tail.